mod_wasm/modules/wasm/mod_wasm.c

```diff
@@ -437,105 +361,77 @@ static int content_handler(request_rec *r)
 }
 
 
-#define WASM_DIRECTIVE_WASMROOT   "WasmRoot"
-#define WASM_DIRECTIVE_WASMMODULE "WasmModule"
-#define WASM_DIRECTIVE_WASMARG    "WasmArg"
-#define WASM_DIRECTIVE_WASMENV    "WasmEnv"
-#define WASM_DIRECTIVE_WASMDIR    "WasmDir"
-#define WASM_DIRECTIVE_WASMMAPDIR "WasmMapDir"
-#define WASM_DIRECTIVE_ENABLECGI  "WasmEnableCGI"
-
-
-static const char *wasm_directive_WasmRoot(cmd_parms *cmd, void *mconfig, const char *word1)
-{
-    x_cfg *cfg = (x_cfg *) mconfig;
-    wasm_config_set_root(word1);
-    return NULL;
-}
-
+#define WASM_DIRECTIVE_WASMMODULE     "WasmModule"
+#define WASM_DIRECTIVE_WASMARG        "WasmArg"
+#define WASM_DIRECTIVE_WASMENV        "WasmEnv"
+#define WASM_DIRECTIVE_WASMDIR        "WasmDir"
+#define WASM_DIRECTIVE_WASMMAPDIR     "WasmMapDir"
+#define WASM_DIRECTIVE_WASMENABLECGI  "WasmEnableCGI"
 
 static const char *wasm_directive_WasmModule(cmd_parms *cmd, void *mconfig, const char *word1)
 {
     x_cfg *cfg = (x_cfg *) mconfig;
-    wasm_config_set_module(word1);
+    wasm_module_load(word1);    // Wasm module is loaded and cached
+    wasm_config_set_module(cfg->loc, word1);    // Wasm config is implictly created for the current location and using the loaded module
     return NULL;
 }
 
 
 static const char *wasm_directive_WasmArg(cmd_parms *cmd, void *mconfig, const char *word1)
 {
     x_cfg *cfg = (x_cfg *) mconfig;
-    if (cfg->configArgCount == CONFIG_DEFINED_ARGS_MAX) {
-      return apr_psprintf(cmd->pool, "WasmArg limit reached in the httpd configuration (maximum is %d)", CONFIG_DEFINED_ARGS_MAX);
-    }
-    configArg *arg = malloc(sizeof(configArg));
-    arg->arg = apr_pstrdup(cmd->pool, word1);
-    cfg->configArgs[cfg->configArgCount] = arg;
-    cfg->configArgCount++;
+    wasm_config_arg_add(cfg->loc, word1);
     return NULL;
 }
 
 
 static const char *wasm_directive_WasmEnv(cmd_parms *cmd, void *mconfig, const char *word1, const char *word2)
 {
     x_cfg *cfg = (x_cfg *) mconfig;
-    if (cfg->configEnvVarCount == CONFIG_DEFINED_ENVVARS_MAX) {
-      return apr_psprintf(cmd->pool, "WasmEnv limit reached in the httpd configuration (maximum is %d)", CONFIG_DEFINED_ENVVARS_MAX);
-    }
-    configEnvVar *env = malloc(sizeof(configEnvVar));
-    env->key = apr_pstrdup(cmd->pool, word1);
-    env->value = apr_pstrdup(cmd->pool, word2);
-    cfg->configEnvVars[cfg->configEnvVarCount] = env;
-    cfg->configEnvVarCount++;
+    wasm_config_env_add(cfg->loc, word1, word2);
     return NULL;
 }
 
 
 static const char *wasm_directive_WasmDir(cmd_parms *cmd, void *mconfig, const char *word1)
 {
     x_cfg *cfg = (x_cfg *) mconfig;
-    wasm_config_add_dir(word1);
+    wasm_config_dir_add(cfg->loc, word1);
     return NULL;
 }
 
 
 static const char *wasm_directive_WasmMapDir(cmd_parms *cmd, void *mconfig, const char *word1, const char *word2)
 {
     x_cfg *cfg = (x_cfg *) mconfig;
-    wasm_config_add_mapdir(word1, word2);
+    wasm_config_mapdir_add(cfg->loc, word1, word2);
```

Comment:
Shouldn't we check the return codes here?

Reply:
I agree